### ISSUE TITLE
Enable 'exec' on /dev/shm, without --privileged

### ISSUE THREAD

**Issue by d11wtq** [docker/docker#6758]
I *think* this used to be possible in the old LXC driver, but under the native driver /dev/shm is mounted with noexec and there doesn't seem to be any way to mount it 'exec' far from running the entire container with `--privileged`. Any thoughts? Some programs require it.

**Comment by jfrazelle** [comment]
you can use `--device` tested locally and it works

**Comment by kkochubey1** [comment]
Hi, How to do it with --device? Could you please show example

**Comment by pwaller** [comment]
Please reopen. @jfrazelle as @kkochubey1, it's not clear how to do that:

```
$ docker run --privileged --device=/dev/shm debian bash
Error response from daemon: Cannot start container f34ae3f5a6534856c95e5aa03eaba808d7fe3b28c05e9f1d703ac6340fdfda53: error gathering device information while adding custom device "/dev/shm": not a device node
```

**Comment by pwaller** [comment]
My need is to remove the `noexec` bit so I can run usermode linux.

**Comment by evgkul** [comment]
If somebody has the same problem, it can be fixed by replacing "/dev/shm" with, for example, "/dev/uml" in /arch/um/os-Linux/mem.c and running "sudo mkdir /dev/uml && sudo chmod 777 /dev/uml" in container before starting usermode linux

**Comment by ligc** [comment]
FYI, 

I am able to run User Mode Linux in docker containers by launching the docker instance with the -v /dev/shm:/dev/shm, as shows below.

docker run -v /dev/shm:/dev/shm  -d <image_name>

**Comment by evgkul** [comment]
ligc,
It's unsafe

**Comment by gvenzl** [comment]
Users face similar issues when running Oracle Database with Java Stored Procedures which happen to need the `exec` flag set for /dev/shm too. I don't believe that `--privileged` is a viable solution around this but that ideally /dev/shm should be mounted with the `exec` flag set.

Can we get more comments around this issue? I don't really like the idea of proposing workarounds which are not safe.

**Comment by kkochubey1** [comment]
@gvenzl on new docker (>1.12.2) there is --shm-size="1g" option. Also you can look into default capbilities or try allow all caps with option --cap-add=ALL

**Comment by gvenzl** [comment]
Thanks @kkochubey1, I know of `--shm-size`, it's been there since a while and we actively use it to size /dev/shm appropriately. But that's the problem, just by specifying `--shm-size` the container gets a mounted /dev/shm with the `noexec` flag set. Even running with `--cap-add=ALL` this doesn't change:

    [root@528b368c8ef0 /]# mount | grep shm
    shm on /dev/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,size=1048576k)

What I'm after is to have /dev/shm mounted without the `noexec`  flag or understand the reason why it is mounted with that flag.

**Comment by gvenzl** [comment]
Can this issue be reopened please until we have further clarification on why this is or isn't possible?

**Comment by rvernica** [comment]
@thaJeztah would you mind re-opening this?

**Comment by rvernica** [comment]
in `1.13.0`, neither `--privileged` not `--cap-add=ALL` will turn off `noexec` on `/dev/shm`:
```bash
> docker run -ti --rm debian mount | grep shm
shm on /dev/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,seclabel,size=65536k)
> docker run -ti --rm --privileged debian mount | grep shm
shm on /dev/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,seclabel,size=65536k)
> docker run -ti --rm --privileged --cap-add=ALL debian mount | grep shm
shm on /dev/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,seclabel,size=65536k)
> docker run -ti --rm --cap-add=ALL debian mount | grep shm
shm on /dev/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,seclabel,size=65536k)
```

**Comment by justincormack** [comment]
You can just mount a tmpfs with exec on /dev/shm with --tmofs I think to
provide different values.

There are security reasons for noexec, and it is I think normal in Linux
distros?

**Comment by rvernica** [comment]
Thanks, I tried this:
```bash
> docker run -ti --rm --tmpfs /dev/shm:rw,nosuid,nodev,size=8g debian
docker: Error response from daemon: linux mounts: Duplicate mount point '/dev/shm'.
```
Do you know how to avoid the *duplicate mount point* situation?

**Comment by thaJeztah** [comment]
hm, that looks like a bug; probably related to https://github.com/docker/docker/pull/29563

**Comment by Angelinsky7** [comment]
i also interested to know how it would be possible to do that without the privileged setting

**Comment by gvenzl** [comment]
@jessfraz, given that many people are interested in that feature and your explanation wasn't clear to at least some of us, can you please reopen this issue and keep tracking it?

**Comment by pannzh** [comment]
Well, I've figured it out finally:
docker run -it -v /dev/shm --tmpfs /dev/shm:rw,nosuid,nodev,exec,size=4g <img> <cmd>

**Comment by Angelinsky7** [comment]
thanks !!!

**Comment by deepeeess** [comment]
@pannzh when I run the command you specified I get a tmpfs of 64M

```
> docker run -v /dev/shm --tmpfs /dev/shm:rw,nosuid,nodev,exec,size=4g --network=nginx-proxy -i -t c73893c2a697 /bin/bash
root@62a22534bbd5:/code# df -h
Filesystem      Size  Used Avail Use% Mounted on
overlay          63G   19G   42G  31% /
tmpfs            64M     0   64M   0% /dev
tmpfs          1000M     0 1000M   0% /sys/fs/cgroup
/dev/vda1        63G   19G   42G  31% /etc/hosts
tmpfs            64M     0   64M   0% /dev/shm
tmpfs          1000M     0 1000M   0% /sys/firmware
root@62a22534bbd5:/code#
```

**Comment by delissonjunio** [comment]
@thaJeztah Can we reopen this? Or at least get some info about the usage of `--device`? As of right now, with the most recent release from edge, I get this:

    > docker run -ti --rm --tmpfs /dev/shm:rw,nosuid,nodev,size=8g debian
    docker: Error response from daemon: linux mounts: Duplicate mount point '/dev/shm'.

I need to mount /dev/shm with both a bigger size and exec flag, and doesn't matter how I play with `--shm-size` or `--tmpfs`, I always get this error.

**Comment by thaJeztah** [comment]
@delissonjunio see https://github.com/moby/moby/pull/35316

**Comment by kolyshkin** [comment]
> docker run -ti --rm --tmpfs /dev/shm:rw,nosuid,nodev,size=8g debian
> docker: Error response from daemon: linux mounts: Duplicate mount point '/dev/shm'.

This might be a different bug :( Let me take a look...

**Comment by thaJeztah** [comment]
Oh; wait this is with `--tmpfs` - yes, looks different (sorry, didn't look right)

**Comment by kolyshkin** [comment]
I was able to reproduce it with dockerd-17.03.0-ce. @delissonjunio can you please file a separate issue for that?

A workaround, I guess (and I haven't checked), is to create a mount in the host system and then bind-mount it to inside the container.

**Comment by thaJeztah** [comment]
The duplicate mount point problem is now fixed on master through https://github.com/moby/moby/pull/35467

**Comment by dlitz** [comment]
What are these alleged security issues with /dev/shm exec?  Are they documented anywhere?  I googled for it, and all I see is other people saying it's *sometimes* mounted noexec "for security" without giving any details.

I'm beginning to suspect that this 'security' issue is an urban legend.  In theory, it could be part of making the entire filesystem W^X, but that's not currently the case (an attacker can just write their executable to /tmp) and it probably never should be by default, so what's the point of making just /dev/shm noexec?

**Comment by kolyshkin** [comment]
@dlitz AFAIK people use `noexec` for `/tmp` and such to avoid exploiting bugs in various web applications like forums etc. Basically, this is a way to protect against executing arbitrary files uploaded by web app users (say a user uploads an "image.jpg" file and there is a bug allowing to run binaries, so they do `chmod u+x` on the "image" and run it).

**Comment by allenyllee** [comment]
> Well, I've figured it out finally:
> 
> docker run -it -v /dev/shm --tmpfs /dev/shm:rw,nosuid,nodev,exec,size=4g <img> <cmd>
> 
>  

 _Originally posted by @evilpan in [#6758](https://github.com/moby/moby/issues/6758#issuecomment-284923933)_

no need to add `-v /dev/shm` option, only need `--tmpfs /dev/shm:rw,nosuid,nodev,exec,size=4g`

compare:

without any options:
```shell
docker run -it ubuntu bash -c "mount | grep /dev/shm"
shm on /dev/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,size=65536k,inode64)
```

with `--tmpfs /dev/shm:rw,nosuid,nodev,exec,size=64m` options:
```shell
docker run -it --tmpfs /dev/shm:rw,nosuid,nodev,exec,size=64m ubuntu bash -c "mount | grep /dev/shm"
tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev,relatime,size=65536k,inode64)

```

alos see: [Control SHM mount options - General - Docker Community Forums](https://forums.docker.com/t/control-shm-mount-options/117631)
